### PR TITLE
[GEP-20] Allow adding HA alpha annotation where none existed

### LINKED ISSUE #6532: Change validation of `alpha.control-plane.shoot.gardener.cloud/high-availability` annotation
**How to categorize this issue?**

/area backup
/area high-availability
/kind enhancement

**What would you like to be added**:
Gardener has an annotation `alpha.control-plane.shoot.gardener.cloud/high-availability` intended to indicate whether the shoot uses a single member etcd or a 3 member etcd cluster
As of now, this annotation is [immutable](https://github.com/shreyas-s-rao/gardener/blob/9a39dd837c84022d022e1362fce0ab37eba1037b/pkg/apis/core/validation/shoot.go#L1800-L1810) so that the cluster size does not change after it is created. This was necessary as etcd-druid only supported multi-node etcd cluster bootstraping
It is now needed that we be allowed to add this annotation when it is not present so that etcd clusters can be scaled up from single member clusters to multi member clusters. This is required as etcd-druid now supports etcd cluster scale up and with the existing annotation validation we cannot take advantage of this feature

**Why is this needed**:
From `etcd-druid v0.12.3` onwards, we now support etcd cluster scale up from a single member etcd to a multi node etcd cluster.
With the current annotation validation, etcd cluster sizes can not be updated to take advantage of this feature. With this requested change, gardener should allow the addition of this annotation so that we can update the replica number only from 1 -> 3 and not the other way around

### PR DESCRIPTION
**Pull request by unmarshall**
**How to categorize this PR?**

/area control-plane
/area backup
/area high-availability
/kind enhancement

**What this PR does / why we need it**:
It is now allowed to add alpha HA annotation to the shoot spec where none existed. Changing the value of an existing HA annotation will still be disallowed as we do not support upgrading or downgrading from a single-zone to multi-zone HA shoot control plane.
In addition it restricts scheduling of HA shoot onto a non multi-zonal seed via admission plugin.

This change will not conflict upcoming changes in [PR#6530](https://github.com/gardener/gardener/pull/6530)

**Which issue(s) this PR fixes**:
Fixes #6532 

**Special notes for your reviewer**:

**Release note**:

```other operator
Adding an alpha HA annotation to the shoot spec where none existed is now allowed.
```

### CODE REVIEW

**Review by timuthy — COMMENTED** [review]
Thanks for the PR 👍 The changes look good. I wondered if you have tested this upgrade scenario (at least in your local setup)?

**Comment by timuthy** [comment]
/hold
- testing needs to be performed
- discussion is needed if further GEP implementations should be done first

**Comment by unmarshall** [comment]
Made some additional changes:
Always configure PeerURLTLS so that even for a single node etcd peer URL uses https. This allows scale up from 1 to 3. New nodes will attempt to reach `etcd-main-0` using https and will fail with:
```
2022-08-23 07:26:56.595765 W | etcdserver: could not get cluster response from https://etcd-main-0.etcd-main-peer.shoot--dev--test1.svc:2380/: Get https://etcd-main-0.etcd-main-peer.shoot--dev--test1.svc:2380/members: http: server gave HTTP response to HTTPS client
```
With this change even the single node etcd will use https protocol for peer URL enabling other members (when scaled) to communicate to it to be added as learner and later on to be promoted as member(s).

**Impact:**
1. For existing clusters which already have a single etcd scale up will **not be** possible out of the box.
2. For new clusters with single node etcd, scale up to 3 would be possible without any manual intervention.

To have scale-up for the first case following needs to be done:
* Exec into the etcd-main-0 and get the member-id
*  Update the peerURL using etcdctl to use https instead of http
*  Restart all members

**Tests**
Using local setup, on a shoot cluster a multi-zonal seed was setup. 
* Created a shoot without `alpha.control-plane.shoot.gardener.cloud/high-availability` annotation. This starts a non-HA shoot cluster.
* Updated the shoot and added annotation `alpha.control-plane.shoot.gardener.cloud/high-availability: multi-zone`. Gardenlet reconciled the cluster and started 3 instances of etcd-main and etcd-event pods successfully.

**Comment by timuthy** [comment]
> To have scale-up for the first case following needs to be done:

    Exec into the etcd-main-0 and get the member-id
    Update the peerURL using etcdctl to use https instead of http
    Restart all members

This definitely sounds like a job for `etcd-druid` and `etcd-backup-restore`.

**Comment by unmarshall** [comment]
> This definitely sounds like a job for etcd-druid and etcd-backup-restore.

Actually there is already code in etcdbr which updates the member peer URL for the case where the etcd replicas = 1 ([updateMember method call](https://github.com/gardener/etcd-backup-restore/blob/master/pkg/server/backuprestoreserver.go#L212-L226)). This is done during the startup of the `etcdbackupserver`.  However for update of the `peerURL` to take effect, a restart of the etcd member is required (see [here](https://etcd.io/docs/v3.3/op-guide/runtime-configuration/#update-advertise-peer-urls)).

Once this PR is merged then from that point onwards,  etcd config will have peer URL TLS enabled. Assuming that all the existing clusters hibernate or get updated regularly, the first restart will ensure that the etcd-backup-restore updates the member URL and the second restart onwards, the changed URL will be in effect. Assuming that hibernation is a periodic activity - all it will take it 2 restarts (could be due to wake-ups from hibernation or a rolling update) and the peer URL will auto-correct.

After this time if HA annotation is added then it will properly scale etcd cluster from 1->3.  So no change will be required in etcd-druid and etcd-backup-restore (change is already in).

Will this be acceptable?

**Comment by timuthy** [comment]
> Once this PR is merged then from that point onwards, etcd config will have peer URL TLS enabled. Assuming that all the existing clusters hibernate or get updated regularly, the first restart will ensure that the etcd-backup-restore updates the member URL and the second restart onwards, the changed URL will be in effect. Assuming that hibernation is a periodic activity - all it will take it 2 restarts (could be due to wake-ups from hibernation or a rolling update) and the peer URL will auto-correct.

IIUC, we should merge this PR but changing an existing cluster to HA will only work if one at **least hibernated their cluster once** of anything else **let to a restart at least once before HA is enabled**?

Btw. doesn't the update to peer-tls alone lead to a restart (w/o enabling HA)?

**Comment by unmarshall** [comment]
Additional change added: Admission plugin now restricts the following:
HA shoot cannot be scheduled onto a non multi-zonal seed.

File: pkg/apis/core/validation/shoot_test.go
Comment (plkokanov) [inline pkg/apis/core/validation/shoot_test.go]:
```suggestion
			It("should allow to set the HAControlPlanes options if it did not exist before", func() {
```
To make it more aligned with the description of the tests above.

File: plugin/pkg/shoot/validator/admission_test.go
Comment (plkokanov) [inline plugin/pkg/shoot/validator/admission_test.go]: I guess this got deleted by accident.

File: plugin/pkg/shoot/validator/admission_test.go
Comment (plkokanov) [inline plugin/pkg/shoot/validator/admission_test.go]:
```suggestion
			DeferCleanup(test.WithFeatureGate(schedulerfeatures.FeatureGate, features.HAControlPlanes, true))
```

Also wouldn't it be better to put this line in a `BeforeEach()` in the `Context("HA shoot scheduling checks on seed", func() { ....` block?

File: plugin/pkg/shoot/validator/admission_test.go
Comment (plkokanov) [inline plugin/pkg/shoot/validator/admission_test.go]: Nit: separate `It()`s with an empty line:
```suggestion

					It("should reject scheduling of single-zonal HA shoot on non multi-zonal seeds", func() {
```

File: plugin/pkg/shoot/validator/admission.go
Comment (plkokanov) [inline plugin/pkg/shoot/validator/admission.go]: What about when the shoot is being rescheduled to a new seed? You can check/modify the scheduling validation here (for the `shoots/binding` subresource): https://github.com/gardener/gardener/blob/038efd2f86d8fb0efd7449990353011595e5a718/plugin/pkg/shoot/binding/admission.go#L211

File: plugin/pkg/shoot/validator/admission.go
Comment (plkokanov) [inline plugin/pkg/shoot/validator/admission.go]:
```suggestion
// TODO(unmarshall): This restriction will be removed in the future.
```

File: pkg/operation/botanist/component/etcd/etcd.go
Comment (plkokanov) [inline pkg/operation/botanist/component/etcd/etcd.go]: Will this change initiate etcd restarts when the corresponding gardener release gets deployed?

File: plugin/pkg/shoot/validator/admission_test.go
Comment (plkokanov) [inline plugin/pkg/shoot/validator/admission_test.go]: This seems to be unused.

**Review by plkokanov — CHANGES_REQUESTED** [review]

File: plugin/pkg/shoot/validator/admission.go
Comment (plkokanov) [inline plugin/pkg/shoot/validator/admission.go]: I think you should use:
```suggestion
		if utilfeature.DefaultFeatureGate.Enabled(features.HAControlPlanes) {
```
Where `utilfeatures` is `utilfeature "k8s.io/apiserver/pkg/util/feature"`

File: plugin/pkg/shoot/validator/admission_test.go
Comment (plkokanov) [inline plugin/pkg/shoot/validator/admission_test.go]: Why is this needed? Additionally `gardenletfeatures` should be used for `gardenlet` related flags.

**Review by plkokanov — CHANGES_REQUESTED** [review]

File: plugin/pkg/shoot/validator/admission_test.go
Comment (unmarshall) [reply plugin/pkg/shoot/validator/admission_test.go]: Feature gate usage was not required and also did not make much sense for admission plugin as no one explicitly adds any feature gates for plugins today. I have removed that check entirely and also changed the unit tests to remove `test.WithFeatureGate`

File: plugin/pkg/shoot/validator/admission_test.go
Comment (unmarshall) [reply plugin/pkg/shoot/validator/admission_test.go]: Ah let me check.

File: plugin/pkg/shoot/validator/admission_test.go
Comment (unmarshall) [reply plugin/pkg/shoot/validator/admission_test.go]: yes i have removed this code.

File: plugin/pkg/shoot/validator/admission.go
Comment (unmarshall) [reply plugin/pkg/shoot/validator/admission.go]: This has now been removed completely.

File: plugin/pkg/shoot/validator/admission_test.go
Comment (unmarshall) [reply plugin/pkg/shoot/validator/admission_test.go]: This has also been removed now.

File: pkg/operation/botanist/component/etcd/etcd.go
Comment (unmarshall) [reply pkg/operation/botanist/component/etcd/etcd.go]: If Peer TLS is changed (enabled) for existing etcd cluster then it will result in 2 restarts of the etcd cluster. This is unfortunately a requirement for etcd. Also See https://github.com/gardener/etcd-druid/issues/416
But this will only trigger it during the maintenance window as by default etcd-druid reconciliation is disabled unless there are previous failures (error was returned during a previous reconcile) or explicit annotation to reconcile has been added.

File: plugin/pkg/shoot/validator/admission_test.go
Comment (plkokanov) [reply plugin/pkg/shoot/validator/admission_test.go]: I guess it makes sense that the check for the HA feature gate is not really needed. 
But just a note that we still use the `SeedChange` feature gate here: https://github.com/gardener/gardener/blob/b01fd6084bb89a42a45ce7f76e20cbd6f0fd17b0/plugin/pkg/shoot/binding/admission.go#L177

**Comment by timuthy** [comment]
@unmarshall this PR is dependent on https://github.com/gardener/etcd-druid/issues/420, right?

**Comment by unmarshall** [comment]
> @unmarshall this PR is dependent on https://github.com/gardener/etcd-druid/issues/420, right?

Yes. I will be shortly filing the PR for etcd issue as well.

File: plugin/pkg/shoot/validator/admission.go
Comment (plkokanov) [inline plugin/pkg/shoot/validator/admission.go]: This validation should be performed if `shootIsBeingScheduled` or `shootIsBeingRescheduled` are true, currently it is only performed when `shootIsBeingScheduled` is true.

@shafeeqes do you think it is ok to keep it only here. It doesn't make too much sense to also add `validateSeedSelectionForHAShoot()` to the `binding` admission plugin

File: plugin/pkg/shoot/validator/admission_test.go
Comment (plkokanov) [inline plugin/pkg/shoot/validator/admission_test.go]: These four lines could be extracted in a `BeforeEach()` call so that they are not repeated for every test, right?

**Review by plkokanov — CHANGES_REQUESTED** [review]

File: plugin/pkg/shoot/validator/admission.go
Comment (shafeeqes) [reply plugin/pkg/shoot/validator/admission.go]: I did not check the PR thoroughly but if it is validating seed, I think it should be there in case of rescheduling as well, ie; in `binding` admission plugin.

Comment (shafeeqes) [reply plugin/pkg/shoot/validator/admission.go]: As Plamen suggested, instead of duplicating, you can keep it in Shoot Validator. but check for both scheduling and rescheduling.

Comment (unmarshall) [reply plugin/pkg/shoot/validator/admission.go]: done

Comment (unmarshall) [reply plugin/pkg/shoot/validator/admission.go]: this change was moved to #6530 and was merged recently. The check is changed.

File: plugin/pkg/shoot/validator/admission_test.go
Comment (unmarshall) [reply plugin/pkg/shoot/validator/admission_test.go]: This change was included in #6530 and relevant tests were written there. Changes were merge to master

**Comment by rfranzke** [comment]
This PR still has the "hold" label. Is it ready for review/what's the status? Thanks!

**Comment by unmarshall** [comment]
>This PR still has the "hold" label. Is it ready for review/what's the status? Thanks!

[etcd-druid:PR#421](https://github.com/gardener/etcd-druid/pull/421) and [etcd-backup-restore: PR#534](https://github.com/gardener/etcd-backup-restore/pull/534) should be merged before this PR gets merged. Having said that, all changes to this PR have been done. So it is open for review but it cannot be merged.

**Comment by rfranzke** [comment]
> > This PR still has the "hold" label. Is it ready for review/what's the status? Thanks!
> 
> [etcd-druid:PR#421](https://github.com/gardener/etcd-druid/pull/421) and [etcd-backup-restore: PR#534](https://github.com/gardener/etcd-backup-restore/pull/534) should be merged before this PR gets merged. Having said that, all changes to this PR have been done. So it is open for review but it cannot be merged.

Then let's convert it to 'draft' for now.

**Comment by unmarshall** [comment]
**NOTE for the reviewers:**
vulnerability check is failing because of https://pkg.go.dev/vuln/GO-2022-0969 which will require `controller-runtime` to move to go version 1.19.1

**Comment by timuthy** [comment]
Thanks for updating this PR. I'll continue with the review and test some cases.

File: pkg/apis/core/validation/shoot.go
Comment (rfranzke) [inline pkg/apis/core/validation/shoot.go]:
```suggestion
			oldShoot.Spec.ControlPlane.HighAvailability.FailureTolerance.Type, "switching from spec.controlPlane.highAvailability.failureTolerance.type to HA annotation is not allowed"))
```

File: pkg/apis/core/validation/shoot.go
Comment (rfranzke) [inline pkg/apis/core/validation/shoot.go]:
```suggestion
					oldAnnotationVal, "forbidden switch from HA annotation to spec.controlPlane.highAvailability.failureTolerance.type, cannot change failure tolerance"))
```

File: pkg/apis/core/validation/shoot.go
Comment (rfranzke) [inline pkg/apis/core/validation/shoot.go]:
```suggestion
				allErrs = append(allErrs, field.Forbidden(field.NewPath("metadata", "annotations").Key(v1beta1constants.ShootAlphaControlPlaneHighAvailability),
```

File: pkg/operation/botanist/component/etcd/etcd.go
Comment (rfranzke) [inline pkg/operation/botanist/component/etcd/etcd.go:501]: Why was this removed?

File: plugin/pkg/shoot/validator/admission_test.go
Comment (rfranzke) [inline plugin/pkg/shoot/validator/admission_test.go]: Can you please revert the changes in this file? They don't seem to be related to the PR.

File: pkg/operation/botanist/component/etcd/etcd.go
Comment (timuthy) [reply pkg/operation/botanist/component/etcd/etcd.go:501]: I'm not sure if it's a good idea to enable peer TLS for all existing shoot clusters even for those which don't request HA. There is a certain risk that re-creating the affected `StatefulSet` two times in a row en masse creates issues as many detach and attach operations are involved. If we want to turn on peerTLS by default also for single-node etcds, then I'd only do it for new clusters.

**Review by timuthy — COMMENTED** [review]
Only one concern for my side. The tests went very well 👍

File: pkg/operation/botanist/component/etcd/etcd.go
Comment (unmarshall) [reply pkg/operation/botanist/component/etcd/etcd.go:501]: Enabling Peer TLS allows easy scale up without any disruption to etcd. For any new single node etcd cluster if Peer TLS is enabled then at a later time if this is upgraded to a HA cluster with 3 replicas for etcd then one just needs to update the StatefulSet replicas and with zero down time the etcd cluster will be updated.

On the other hand if TLS is disabled for any single node cluster and now there is an attempt to scale it up to 3 then it will require 2 restarts of the existing etcd member (to ensure that the peer URL TLS enablement is reflected in both etcd DB and the running etcd process) before the cluster can be scaled to 3 causing downtime.

If there is no additional penalty that is paid for enabling TLS then IMHO it should always be enabled.

>If we want to turn on peerTLS by default also for single-node etcds, then I'd only do it for new clusters.

Ok that is also fine with me.

File: plugin/pkg/shoot/validator/admission_test.go
Comment (unmarshall) [reply plugin/pkg/shoot/validator/admission_test.go]: Sure. Initially there were changes to this file and later those were reverted back. I will revert all changes to this file.

Comment (unmarshall) [reply plugin/pkg/shoot/validator/admission_test.go]: done

File: pkg/operation/botanist/component/etcd/etcd.go
Comment (unmarshall) [reply pkg/operation/botanist/component/etcd/etcd.go:501]: made the change. PTAL

Comment (timuthy) [reply pkg/operation/botanist/component/etcd/etcd.go:501]: As discussed with @rfranzke and @unmarshall, we plan to rollout PeerTLS to all clusters because it brings several benefits:
- Zero downtime upgrades to HA
- Consistent landscape
- Simpler code

We anyway rollout a new `etcd-backup-restore` version in the next Gardener release which will result in etcd restarts. So it'd be perfect if we combine this with the PeerTLS activation.

Comment (unmarshall) [reply pkg/operation/botanist/component/etcd/etcd.go:501]: Ok done the changes. PTAL
Cleaned up the commits and squashed it into 1 commit.

**Review by timuthy — COMMENTED** [review]
I tested the PR but noticed that etcd pods restarted more often than the expected because of Druid's instructions:

```
2022-09-20T14:35:38.011Z	INFO	etcd-controller	added tasks to deploy flow due to peer url tls changed to enabled	{"etcd": "shoot--dev--d066080aws-fsttm/etcd-events", "operation": "reconcile", "sts": "shoot--dev--d066080aws-fsttm/etcd-events", "namespace": "shoot--dev--d066080aws-fsttm", "name": "etcd-events", "etcdUID": "f5b4af82-d8c6-47aa-8a16-9d428464d3d2", "replicas": 1}
2022-09-20T14:35:38.041Z	INFO	etcd-controller	added tasks to deploy flow due to peer url tls changed to enabled	{"etcd": "shoot--dev--d066080aws-fsttm/etcd-main", "operation": "reconcile", "sts": "shoot--dev--d066080aws-fsttm/etcd-main", "namespace": "shoot--dev--d066080aws-fsttm", "name": "etcd-main", "etcdUID": "f822ec13-f1d4-4047-8cd7-844cf106926c", "replicas": 1}
2022-09-20T14:36:53.229Z	INFO	etcd-controller	added tasks to deploy flow due to peer url tls changed to enabled	{"etcd": "shoot--dev--d066080aws-fsttm/etcd-events", "operation": "reconcile", "sts": "shoot--dev--d066080aws-fsttm/etcd-events", "namespace": "shoot--dev--d066080aws-fsttm", "name": "etcd-events", "etcdUID": "f5b4af82-d8c6-47aa-8a16-9d428464d3d2", "replicas": 1}
```

It seems like especially the `etcd-events` resources was reconciled for a second time too fast and the second time PeerTLS was still not considered active.

**Review by rfranzke — COMMENTED** [review]
I don't have further comments @timuthy :)
/lgtm

**Comment by timuthy** [comment]
Thanks @rfranzke. Due to https://github.com/gardener/gardener/pull/6533#pullrequestreview-1113981165 @unmarshall and I decided to just enable Peer-TLS for new cluster or when HA is enabled. After issue https://github.com/gardener/etcd-backup-restore/issues/538 is resolved, we can also enable it for all existing clusters.

**Comment by rfranzke** [comment]
Well, should we just wait for https://github.com/gardener/etcd-backup-restore/issues/538 and remove this peer certificate change from this PR then?

**Comment by timuthy** [comment]
> Well, should we just wait for [gardener/etcd-backup-restore#538](https://github.com/gardener/etcd-backup-restore/issues/538) and remove this peer certificate change from this PR then?

Okay, changes are removed 🙂

**Pull request merged**
